### PR TITLE
Return a deterministic trace generator when using HMC with parallel chains

### LINKED ISSUE #1505: HMC gives non deterministic results with parallel chaining
Even though the `ParallelSampler` yields the same set of traces (each worker gives deterministic results), the order in which the traces are consumed might be different, which gives different results when we use `TracePredictive`. 

One way to get deterministic results from HMC/NUTS with `num_chains > 1` would be to store intermediate results, and yield results from each of the workers in a round-robin fashion.

### PR DESCRIPTION
**Pull request by neerajprad**
Fixes #1505. 

While each of the workers in `mcmc._ParallelSampler` returns deterministic outputs, the order of the traces returned by `_ParallelSampler._traces` generator is non-deterministic (it depends on how fast/slow each of the workers is). This gives deterministic stats on the latents, but non-deterministic results when the `TracePosterior` object is consumed by `TracePredictive`.

As a solution, we are yielding traces from each of the chains in a round-robin fashion and using a queue to hold the intermediate traces from the workers. 

Note that this is not a final interface and may need a slight refactoring (or more design discussion) when we would like to compute and provide diagnostics that need chain information. In that case, we may have to provide an optional chain id to the parent `TracePosterior` class.

### CODE REVIEW

File: pyro/infer/mcmc/mcmc.py [hunk 1]
```diff
@@ -162,10 +167,23 @@ def _traces(self, *args, **kwargs):
...
                     raise val
-                elif val is not None:
-                    yield val
-                else:
+                results_buffer[chain_id - 1].append(val)
+                if val is None:
                     active_workers -= 1
+                while results_buffer[cur_idx]:
+                    val = results_buffer[cur_idx].popleft()
+                    if val:
```
Comment (fritzo) [inline pyro/infer/mcmc/mcmc.py]: what does `if val` mean here? I.e. what types and values can `val` take on? Can you more precisely test `if val is not None`?

File: pyro/infer/mcmc/mcmc.py [hunk 2]
```diff
@@ -162,10 +167,23 @@ def _traces(self, *args, **kwargs):
                 if isinstance(val, Exception):
                     # Exception trace is already logged by worker.
                     raise val
-                elif val is not None:
-                    yield val
-                else:
+                results_buffer[chain_id - 1].append(val)
```
Comment (fritzo) [inline pyro/infer/mcmc/mcmc.py]: It seems a little weird to `.append(None)` and then not yield it. Would it be simpler to rearrange so that you
```py
if val is None:
    active_worker -= 1
else:
    results_buffer[chain_id - 1].append(val)
    ...
```

File: pyro/infer/mcmc/mcmc.py [hunk 3]
```diff
@@ -144,6 +144,11 @@ def _traces(self, *args, **kwargs):
         # restore original handler
         signal.signal(signal.SIGINT, sigint_handler)
         active_workers = self.num_chains
+        # To yield a deterministic ordering, we hold intermediate traces
+        # from each of the workers in its own queue in `results_buffer`
+        # and yield these in a round robin fashion.
+        cur_idx = 0
```
Comment (fritzo) [inline pyro/infer/mcmc/mcmc.py]: nit: how about `current_chain` or `chain_idx`? I was a bit confused whether this indexed over workers or chains or time

**Comment by neerajprad** [comment]
> Outside of this PR, I'm concerned that we're spin locking rather than blocking. Why not block?

Unfortunately, the issue with a blocking `queue.get` is that it also disregards KeyboardInterrupt, so will lead to situations where the user will need to kill off the processes manually to recover. We can however specify a small timeout, say `0.1s` so as to catch a KeyboardInterrupt - I am not doing that since the main process is pretty lightweight anyways, it just handles logging and aggregating traces from the workers and exits the loop as soon as any of the worker crashes or there is a KeyboardInterrupt.

File: pyro/infer/mcmc/mcmc.py (hunk 3 above)
Comment (neerajprad) [reply pyro/infer/mcmc/mcmc.py]: This is the round robin index, i.e. the index of the queue in the buffer we are currently blocked on and will be incremented with each `yield`. This might not be the same as the currently sampled chain, which is given by `chain_id`. Maybe we can call this `buffer_idx`?

File: pyro/infer/mcmc/mcmc.py (hunk 2 above)
Comment (neerajprad) [reply pyro/infer/mcmc/mcmc.py]: Thanks for the suggestion. The while loop can be outside, and this should work.

File: pyro/infer/mcmc/mcmc.py (hunk 1 above)
Comment (neerajprad) [reply pyro/infer/mcmc/mcmc.py]: This isn't needed - I will remove it.

File: pyro/infer/mcmc/mcmc.py (hunk 3 above)
Comment (fritzo) [reply pyro/infer/mcmc/mcmc.py]: How are these numbers related: number of workers, number of chains, and number of buffers?

**Comment by fritzo** [comment]
> the main process is pretty lightweight anyways

But a process that spin locks still uses a full cpu core, right? But ok if you have 56 cores I guess 1 is lightweight...

File: pyro/infer/mcmc/mcmc.py (hunk 3 above)
Comment (neerajprad) [reply pyro/infer/mcmc/mcmc.py]: They are the same - `chain_id` refers to the id of the chain/worker from where the last trace is sampled. `(chain_id, trace)` is sent as a tuple by each of the workers. As soon as this sample is received it is put into its own queue in the buffer. Then we keep yielding samples from the buffer (popping off the queue indexed by `buffer_idx`, which is incremented) until we are blocked, at which point we call `.get_nowait` on the shared queue again.

**Comment by neerajprad** [comment]
> But a process that spin locks still uses a full cpu core, right? Ok if you have 56 cores I guess 1 is lightweight...

That is true - I added a timeout of 1 s., and it seems to handle keyboard interrupts fine without deadlocking, so let me address it in this PR.

**Comment by neerajprad** [comment]
Thanks for reviewing, @fritzo. There are still some rare cases when the logger thread is not able to clean up on KeyboardInterrupt, but it is not related to changing to a blocking get with timeout, so I will try to address that separately.

**Comment by neerajprad** [comment]
For some reason, this seems to hang on travis, so I will need to debug to see what's happening.

**Comment by fritzo** [comment]
@neerajprad feel free to remove the block-vs-spin change to unblock this PR. It's a minor issue and we can address it another time.

**Comment by neerajprad** [comment]
Thanks, @fritzo. I have reverted the block with timeout change, but will address this in another PR.
